File: sorting/quickSort/quickSort.py
array1 = [9,1,4,6,2,3,6,76,23,1]
array2 = [91,123,23,523,12,5]
array3 = [9,8,7,6,5,4,3,2,1]

def partition(arr, start, end):

    # variable for our pivot to compare other values to inside the array
    pivot = arr[end]

    # indices to move throughout the list from start to end, end to start
    leftIndex = start
    rightIndex = end-1

    # while loop to make sure our indices don't cross
    while leftIndex <= rightIndex:

        # makes sure that our left index doesn't move out of range and that each value we pass is less than our pivot value
        while leftIndex <= end and arr[leftIndex] < pivot:
            
            # increments if those conditions are met
            leftIndex += 1

        # makes sure that our right index doesn't move out of range and that each value we pass is greater than/equal to our pivot value
        while rightIndex >= start and arr[rightIndex] >= pivot:

            # increments if those conditions are met
            rightIndex -= 1

        # if we still have values to consider, that means we've got values at indices that need to be swapped
        if leftIndex < rightIndex:
            arr[leftIndex], arr[rightIndex] = arr[rightIndex], arr[leftIndex]

        # else we need to move the pivot into the correct spot by swapping it with where our left index ended
        else:
            arr[leftIndex], arr[end] = arr[end], arr[leftIndex]

    return leftIndex

def quickSortSubArr(arr, start, end):
    
    # recursive function, so if we have a list of length 0 or 1, return without partitioning further
    if start >= end:
        return

    # divide the array into subarrays with the above function
    pivotIndex = partition(arr, start, end)

    # recursively sort each sublist 
    quickSortSubArr(arr, start, pivotIndex-1)
    quickSortSubArr(arr, pivotIndex+1, end)

def quickSort(arr):
    quickSortSubArr(arr, 0, len(arr)-1)

    return arr


print(quickSort(array1))
print(quickSort(array2))
print(quickSort(array3))

File: sorting/quickSort/quickSort.js
array1 = [9,1,4,6,2,3,6,76,23,1]
array2 = [91,123,23,523,12,5]
array3 = [9,8,7,6,5,4,3,2,1]

function partition(arr, start, end) {

    // start with a pivot variable to compare values in our arr to. To start, with use the last item in the arr
    let pivot = arr[end]

    // indices that will move through the list. Left goes right, right moves left
    let leftIndex = start
    let rightIndex = end-1

    // overarching while loop that makes sure our indices don't pass each other
    while (leftIndex <= rightIndex) {

        // first while loop makes sure that our left index doesn't pass the end of the list, and that value of that index is less than the value stored in our pivot
        while (leftIndex <= end && arr[leftIndex] < pivot) {

            // increments the index if those conditions are met
            leftIndex++
        }

        // this loop makes sure that our right index doesn't go past the start of the list and that the value of that index is greater than/equal to the value stored in our pivot
        while(rightIndex >= start && arr[rightIndex] >= pivot) {

            // decrements the index if those condtions are met
            rightIndex--
        }

        // if we reach these conditions, that means that we need to make a switch
        // if the left index is less than the right index, means there's still sorting to do and we've reached two values that need to be swapped
        if (leftIndex < rightIndex) {

            // use a temp variable to swap the values at those indices
            const temp = arr[leftIndex]
            arr[leftIndex] = arr[rightIndex]
            arr[rightIndex] = temp

        // else we've looked at all the values and are done partitioning.
        // we move the pivot to its correct spot in the list
        } else {
            const temp = arr[leftIndex]
            arr[leftIndex] = arr[end]
            arr[end] = temp
        }
    }

    // returns new pivot index for next step
    return leftIndex
}

function quickSortSubArray(arr, start, end) {

    // recursive function, so this returns if we have a list length of 1
    if (start >= end){
        return
    }

    // divide the array into two arrays
    const pivotIndex = partition(arr, start, end)

    // recurively sort each sub array
    quickSortSubArray(arr, start, pivotIndex-1)
    quickSortSubArray(arr, pivotIndex+1, end)
}

function quickSort(arr) {
    quickSortSubArray(arr, 0, arr.length-1)

    return arr
}


console.log(quickSort(array1))
console.log(quickSort(array2))
console.log(quickSort(array3))